component/status/status.go

```diff
@@ -0,0 +1,194 @@
+// An event can signal that a component is working normally (i.e. Type: status.OK), or that it
+// is in an error state (i.e. Type: status.RecoverableError). An error status may optionally
+// include an error object to provide additional insight to registered listeners.
+type Event struct {
+	eventType   EventType
+	timestamp   uint64
+	componentID config.ComponentID
+	err         error
+}
+
+// Type returns the event type
+func (ev *Event) Type() EventType {
+	return ev.eventType
+}
+
+// Timestamp returns the time of the event in nanos
+func (ev *Event) Timestamp() uint64 {
+	return ev.timestamp
+}
+
+// ComponentID returns the ID of the component that generated the status.Event
+func (ev *Event) ComponentID() config.ComponentID {
+	return ev.componentID
+}
+
+// Err returns the error associated with the status.Event
+func (ev *Event) Err() error {
+	return ev.err
+}
+
+// EventOption applies options to a status.Event
+type EventOption func(*Event) error
+
+// WithComponentID sets the ComponentID that generated the status.Event
+func WithComponentID(componentID config.ComponentID) EventOption {
+	return func(o *Event) error {
+		o.componentID = componentID
+		return nil
+	}
+}
+
+// WithTimestamp sets the timestamp, expected in nanos, for a status.Event
+func WithTimestamp(nanos uint64) EventOption {
+	return func(o *Event) error {
+		o.timestamp = nanos
+		return nil
+	}
+}
+
+// WithError assigns an error object to a Event. It is optional and should only be applied
+// to a Event of type: RecoverableError, PermanentError, or FatalError.
+func WithError(err error) EventOption {
+	return func(o *Event) error {
+		if o.eventType == OK {
+			return errors.New("event with status.OK cannot have an error")
+		}
+		o.err = err
+		return nil
+	}
+}
+
+// NewEvent creates and returns a status.Event with default and / or the provided options. Will
+// return an error if an error is provided for a non-error event type (e.g. status.OK)
+func NewEvent(eventType EventType, options ...EventOption) (*Event, error) {
+	ev := Event{
+		eventType: eventType,
+	}
+
+	for _, opt := range options {
+		if err := opt(&ev); err != nil {
+			return nil, err
+		}
+	}
+
+	if ev.timestamp == 0 {
+		ev.timestamp = uint64(time.Now().UnixNano())
+	}
+
+	return &ev, nil
+}
+
+// EventFunc is a callback function that receives status.Events
+type EventFunc func(event *Event) error
+
+// PipelineFunc is a function to be called when the collector pipeline changes states
+type PipelineFunc func() error
+
+// UnregisterFunc is a function to be called to unregister a component that has previously
+// registered to listen to status notifications
+type UnregisterFunc func() error
+
+var noopStatusEventFunc = func(event *Event) error { return nil }
+
+var noopPipelineFunc = func() error { return nil }
+
+// Listener is a struct that manages handlers to status and pipeline events
+type Listener struct {
```

Comment:
This struct is being used by the notifications implementation in service/internal/components. When the notifications implementation was in component I was able to keep it private. The public helpers are so that it is immutable after construction. Is there a better way to organize the package layout to hide it? I was just wondering if you have anything in mind, no problem if you don't, and I can think more about this if it's important.

Reply:
But this can also be in a internal package correct?

Reply:
For me it is very important to not expose anything public that does not need to be public.

Reply:
There are two packages that depend on the status listener `component` and `service/internal/components`. `host.RegisterStatusListener` takes parameters of type `StatusListenerOption` to used to construct a StatusListener. After construction, the StatusListener is used in `service/internal/components`. If the `StatusListener` was only used in the `components` package, I could see how we could put it in an internal package, but since it's also used in `service/internal/components` this complicates things. I can not see an easy way to move this to an internal package where everything that needs it will have access to it. Let me know if I am missing something or alternatively, if there is a better way to separate the options that are passed to `host.RegisterStatusListener` from the construction of the `StatusListener` as I think that is root cause of this issue.